src/github/credentials.ts

```diff
@@ -3,48 +3,127 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
  *--------------------------------------------------------------------------------------------*/
 
-import { Configuration } from '../configuration';
-import { Remote } from '../common/remote';
+import * as Octokit from '@octokit/rest';
 import { fill } from 'git-credential-node';
-const Octokit = require('@octokit/rest');
+import * as vscode from 'vscode';
+import { IHostConfiguration, HostHelper } from '../authentication/configuration';
+import { GitHubServer } from '../authentication/githubServer';
+import { Remote } from '../common/remote';
+import { VSCodeConfiguration } from '../authentication/vsConfiguration';
+import Logger from '../common/logger';
+
+const SIGNIN_COMMAND = 'Sign in';
 
 export class CredentialStore {
-	private _octokits: { [key: string]: any };
-	private _configuration: Configuration;
-	constructor(configuration: Configuration) {
+	private _octokits: Map<string, Octokit>;
+	private _configuration: VSCodeConfiguration;
+	constructor(configuration: any) {
 		this._configuration = configuration;
-		this._octokits = [];
+		this._octokits = new Map<string, Octokit>();
 	}
 
-	reset() {
-		this._octokits = [];
+	public reset() {
+		this._octokits = new Map<string, Octokit>();
 	}
 
-	async getOctokit(remote: Remote) {
-		if (this._octokits[remote.url]) {
-			return this._octokits[remote.url];
+	public async getOctokit(remote: Remote): Promise<Octokit> {
+		// the remote url might be http[s]/git/ssh but we always go through https for the api
+		// so use a normalized http[s] url regardless of the original protocol
+		const normalizedUri = remote.gitProtocol.normalizeUri();
+		const host = `${normalizedUri.scheme}://${normalizedUri.authority}`;
+
+		// for authentication purposes only the host portion matters
+		if (this._octokits.has(host)) {
+			return this._octokits.get(host);
 		}
 
-		if (this._configuration.host === remote.host && this._configuration.accessToken) {
-			this._octokits[remote.url] = Octokit({});
-			this._octokits[remote.url].authenticate({
-				type: 'token',
-				token: this._configuration.accessToken
-			});
-			return this._octokits[remote.url];
+		this._configuration.setHost(host);
+
+		let octokit: Octokit;
+		const creds: IHostConfiguration = this._configuration;
+		const server = new GitHubServer(host);
+		let error: string;
+
+		if (creds.token && await server.validate(creds.username, creds.token)) {
+			octokit = this.createOctokit('token', creds);
 		} else {
-			const data = await fill(remote.url);
-			if (!data) {
-				return null;
+
+			// see if the system keychain has something we can use
+			const data = await fill(host);
+			if (data) {
+				const login = await server.validate(data.username, data.password);
```

Comment:
I'm using personal access token for https remotes so this line is being executed as I have username/token saved in git credentials. However this line never returns. Once I commented out this part and it fell back to the oauth login, everything works as expected.

Reply:
Interesting, I wonder what's going on. It might be throwing somewhere and getting swallowed. I'll try to repro locally.

Reply:
I *think* I fixed this, there were some very silly bugs in that code.